Docs: add headless mode and Solana batch guides (#2867)
MetaMask/metamask-docs: metamask-connect/solana/guides/batch-transactions.md

---
title: 'Send Batch Solana Transactions - MetaMask Connect'
sidebar_label: Send batch transactions
description: Sign and send multiple Solana transactions in a single wallet interaction using the signAndSendAllTransactions wallet-standard feature.
keywords:
  [
    solana,
    batch transactions,
    signAndSendAllTransactions,
    wallet-standard,
    MetaMask,
    Connect,
    multiple transactions,
  ]
---

# Send batch transactions

> The `solana:signAndSendAllTransactions` wallet-standard feature lets you sign and send multiple

Reviewer: ```suggestion
The `solana:signAndSendAllTransactions` Wallet Standard feature lets you sign and send multiple
```

Solana transactions in a single wallet interaction.
This is useful for operations that span several transactions, such as initializing multiple accounts,
batch token transfers, or multi-step program interactions.

## Prerequisites

Set up a Solana client and connect to the user's wallet:

```javascript
import { createSolanaClient } from '@metamask/connect-solana'
import {
  Connection,
  Transaction,
  SystemProgram,
  PublicKey,
  LAMPORTS_PER_SOL,
} from '@solana/web3.js'

const solanaClient = await createSolanaClient({
  dapp: {

>     name: 'My Solana DApp',

Reviewer: ```suggestion
    name: 'My Solana Dapp',
```

    url: window.location.origin,
  },
})

const wallet = solanaClient.getWallet()
const { accounts } = await wallet.features['standard:connect'].connect()
const account = accounts[0]
const publicKey = new PublicKey(account.address)
const connection = new Connection('https://solana-devnet.infura.io/v3/<YOUR_INFURA_API_KEY>')
```

## Build and send batch transactions

### 1. Build the transactions

> Construct each transaction with a fresh blockhash and fee payer.

Reviewer: ```suggestion
Construct each transaction with a new block hash and fee payer.
```

  chain: 'solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1',
})
```

### 3. Confirm each transaction

Always confirm transactions before updating the UI:

```javascript
for (const { signature } of results) {
  const confirmation = await connection.confirmTransaction(
    {
      signature,
      blockhash,
      lastValidBlockHeight: (await connection.getLatestBlockhash()).lastValidBlockHeight,
    },
    'confirmed'
  )

  if (confirmation.value.err) {
    console.error('Transaction failed on-chain:', confirmation.value.err)
  } else {
    console.log('Transaction confirmed:', signature)
  }
}
```

## Error handling

```javascript
try {
  const results = await batchFeature.signAndSendAllTransactions({
    account,
    transactions: serializedTransactions,
    chain: 'solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1',
  })
} catch (err) {
  if (err.code === 4001) {
    // User rejected the batch — show retry UI
  } else if (err.code === -32002) {
    // Request already pending — ask user to check MetaMask
  } else {
    console.error('Batch transaction error:', err)
  }
}
```

## Important considerations

- **Fresh blockhash:** Blockhashes expire after ~60 seconds.

>   Fetch `getLatestBlockhash()` immediately before building the batch.

Reviewer: ```suggestion
- **New block hash:** Block hashes expire after ~60 seconds.
  Call `getLatestBlockhash` immediately before building the batch.
```
